app/code/Bulbulatory/Recommendations/Controller/Customer/ConfirmRecommendation.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace Bulbulatory\Recommendations\Controller\Customer;
+
+use Exception;
+use Magento\Framework\App\Action\Context;
+use Magento\Framework\App\Action\Action;
+use Bulbulatory\Recommendations\Model\RecommendationRepository;
+use Psr\Log\LoggerInterface;
+
+class ConfirmRecommendation extends Action
+{
+    /**
+     * @var RecommendationRepository
+     */
+    private $recommendationsRepository;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var
+     */
+    private $hash;
+
+    /**
+     * ConfirmRecommendation constructor.
+     * @param Context $context
+     * @param RecommendationRepository $recommendationRepository
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        Context $context,
+        RecommendationRepository $recommendationRepository,
+        LoggerInterface $logger
+    )
+    {
+        $this->recommendationsRepository = $recommendationRepository;
+
+        $this->logger = $logger;
+
+        parent::__construct($context);
+    }
+
+    /**
+     * @return mixed
+     */
+    public function execute()
+    {
+        try {
+            $this->validateHash();
+            $this->recommendationsRepository->confirm($this->hash);
+            $this->messageManager->addSuccessMessage(__('Thank you for visiting to Bulbulatory.pl!'));
+        }
+        catch (\Throwable $exception)
+        {
+            $this->messageManager->addErrorMessage(__($exception->getMessage()));

```

Comment:
We shouldn't display the exception message to the customer as it may expose some internal information. It may be ok in the admin panel as only trusted people have access there but when it comes to customers, it's better to display some generic message like `There was an error while trying to confirm a recommendation` and log the exception as you already did in the next line.

You can display the exception message if you're catching an exception that you're sure it's not exposing anything. I see you're throwing exceptions in the `validateHash()` method so maybe change them to less generic `InvalidArgumentException` and catch this type of exception, then displaying the exception message will be fine.

Reply:
@TM18

> We shouldn't display the exception message to the customer as it may expose some internal information. It may be ok in the admin panel as only trusted people have access there but when it comes to customers, it's better to display some generic message like There was an error while trying to confirm a recommendation and log the exception as you already did in the next line.
> 
> You can display the exception message if you're catching an exception that you're sure it's not exposing anything. I see you're throwing exceptions in the validateHash() method so maybe change them to less generic InvalidArgumentException and catch this type of exception, then displaying the exception message will be fine

Changed!